### ISSUE TITLE
Add PascalCase capitalisation_policy

### ISSUE THREAD

**Issue by panasenco** [sqlfluff/sqlfluff#820]
[L014](https://docs.sqlfluff.com/en/stable/rules.html#sqlfluff.core.rules.std.Rule_L014) (Inconsistent capitalisation of unquoted identifiers) needs a PascalCase capitalisation_policy option, implemented as follows:

- Unquoted identifier does not contain non-alphanumeric characters
- Unquoted identifier begins with a capital letter

**Comment by alanmcruickshank** [comment]
This sounds doable. We have the added benefit that implementing it in L010 will also get the other capitalisation rules too.

I don't know a feasible way of detecting words *within* an identifier so that would mean:
- `PascalCase` would pass
- `Pascalcase` would pass
- `pascalCase` would fail and be corrected to `Pascalcase`
- `PasCalCaSe` would pass
- `PAscalcase` would pass
- `PASCALCASE` would also pass I think. @panasenco do you agree?
- `pascalcase` would fail and be corrected to `Pascalcase`
- `pascal_case` would tail and be corrected to `Pascalcase`

**Comment by barrywhart** [comment]
I agree -- detecting words within an identifier would involve a lot of guesswork by the implementer and/or the SQLFluff runtime. Some potential ideas that might help a bit:
* Embed a human-language dictionary. Cons: 1. Often a portion of a word is also a word, e.g. "Begin" or "BegIn". I suppose you could have it pick the *longest* words whenever there's ambiguity. 2. Many technical words won't be in it, but these tend to be organization specific, so there could be a configuration setting for it.
* There is one Python package and a few articles about algorithms for splitting "words" into syllables. Of course, syllable != word.
** https://pypi.org/project/syllapy/
*** https://medium.com/@mholtzscher/programmatically-counting-syllables-ca760435fab4
** https://stackoverflow.com/questions/405161/detecting-syllables-in-a-word
** https://zacs.site/blog/building-a-syllable-dictionary-with-python.html

Neither of these sounds like a completely robust solution, but the first one might work "well enough" for some users. I don't see this as a good feature for SQLFluff core, but if you or someone is interested in exploring it, it could be implemented as a **plugin**.

[Here](https://github.com/sqlfluff/sqlfluff/blob/master/docs/source/developingplugins.rst) is the documentation on how to develop plugins. This is a new feature, just added to SQLFluff last week. It is not yet available in a release, but I think there will be a release in the next few weeks. If I have time, I hope to start work soon (next week) on a plugin for my job as well.

**Comment by panasenco** [comment]
> 
>     * `PASCALCASE` would also pass I think. @panasenco do you agree?
> 

Agreed, @alanmcruickshank.

> 
>     * `pascal_case` would tail and be corrected to `Pascalcase`
>

I think you meant `pascal_case` would fail and be corrected to `PascalCase`. If so, agreed.

Additionally, `PASCAL_CASE` would fail and be converted to `PascalCase`.

I don't think there's a need to detect words within identifiers, @barrywhart. Sane people will either separate their words with underscores or use PascalCase. Someone who names their identifiers `runonsamecasecolumnname` is beyond saving anyway.

**Comment by WittierDinosaur** [comment]
> > pascal_case would tail and be corrected to Pascalcase
> I disagree and think this should be corrected to Pascal_Case While I understand converting from snake_case to PascalCase could allow this optimisation, this is too dangerous a rule IMHO to implement widely as it changes the actual identifier beyond just case. Also, as I understand it, there is nothing to prevent underscores in PascalCase?
I disagree on this one. I think if we are going to support conversions at all, they should stay true to their definition. I think this should correct to PascalCase.

**Comment by tunetheweb** [comment]
I'm just concerned that we break things by forcing this fix.

Rule 014 (currently the only rule using PascalCasE) is used for unquoted identifiers include, for example table names.

So this:

```sql
select a
from table_a
```

Would get fixed to this if the policy was set to PascalCase:

```sql
select Ab
from TableA
```

That to me risks breaking SQL as presumably the table is actually called `table_a` and not`TableA`? While the case is not necessarily important (at least in most SQL dialects I know) the addition/removal of the underscore definitely is.

**Comment by tunetheweb** [comment]
Similarly:

```sql
select
    a,
    _b
from
    table1
```

would be incorrectly "fixed" to:


```sql
select
    A,
    B
from
    Table1
```

**Comment by WittierDinosaur** [comment]
I see where you're coming from about mutating the meaning of an identifier. My counterpoint would be that PascalCase is a weird choice for SQL anyway, and by doing as above we aren't actually implementing PascalCase.

```
ThisIsGoodPascalCase
THisISBAdPaSCAlCase
But_This_Is_Not_Pascal_Case
```

**Comment by WittierDinosaur** [comment]
Though I suppose our implementations of upper and lowercase aren't necessarily accurate either. As they both allow underscores, I guess this is consistent. It depends I suppose on whether we want there to be a distinction between:

```
UPPERCASE
SCREAMING_SNAKE_CASE
and
lowercase
snake_case
```

If we do, we shouldn't include underscores in Pascal, and if we don't we should. 🤷‍♂️

**Comment by tunetheweb** [comment]
> My counterpoint would be that PascalCase is a weird choice for SQL anyway

Well certainly PascalCase with underscores is definitely a weird choice!

> It depends I suppose on whether we want there to be a distinction between:
> 
> ```
> UPPERCASE
> SCREAMING_SNAKE_CASE
> and
> lowercase
> snake_case
> ```
> 
> If we do, we shouldn't include underscores in Pascal, and if we don't we should. 🤷‍♂️

I think if we supported `snake_case` there'd be an argument to "fix" underscores when switching between cases (though I'd probably still argue not to, unless we 100% knew we were coming from snake_case - which we won't). But since we don't, I think safer to ignore underscores for our rules.

The other alternative is to have another rule (or option for rules L010 and L014) to allow/disallow underscores. But until then I think suggested fix in #1394 is the best.

**Comment by tunetheweb** [comment]
Plus many, many keywords in SQL are "Uppercase with Underscores".
And some dialects treat underscores specially (e.g. BigQuery states "Unquoted identifiers must begin with a letter or an underscore character.")

So underscores do have special meaning in SQL so shouldn't be presumed to be meaningless, like is being discussed here.

**Comment by WittierDinosaur** [comment]
Yeah, okay, I think I'm convinced. But to pull at this thread, if we're treating underscore as valid space separators, should 
`PASCAL_CASE` not become `Pascal_Case` ?

**Comment by tunetheweb** [comment]
Getting there with you! 😁 Good to thrash it out though, to make sure things aren't missed.

`PASCAL_CASE` is arguable as `Pascal_Case`, I agree.
But then why are we saying `PASCALCASE` is OK, but not `PASCAL_CASE` if we're ignoring underscores?

At the moment my Fix does correct `Pascal_case` to `Pascal_Case` but not `PASCAL_CASE` to `Pascal_Case`

**Comment by tunetheweb** [comment]
Ah I see, it depends if we are treating underscores as "valid space separators" or "ignoring". I was gong for the latter rather than the former to be honest. And do like that for simplicity sake too.

**Comment by WittierDinosaur** [comment]
Yeah, fair. I don't have too strong an opinion on it.
